### ISSUE TITLE
Weave state becomes inconsistent due to scaling

### ISSUE THREAD

**Issue by zacblazic** [weaveworks/weave#3677]
## What you expected to happen?

The network should be able to handle scaling events (up and down) with grace and not cause the network control plane to transition to an unrecoverable & inconsistent state.

## What happened?

The cluster was scaled up from 62 to 102 nodes. The scale up was handled well and apart from a few logs that indicated we were in "sleeve" mode and that we had "multiple connections", however those logs only happened during startup of weave on the new nodes.

Next, the cluster was scaled back down to 62 nodes, which is where the trouble began from what I can tell:

![image](https://user-images.githubusercontent.com/310235/62135736-0607b700-b2e3-11e9-8a4a-ea3b6720d05d.png)

During the scale down, weave state become inconsistent. These are the symptoms observed, some of which can be seen on the charts above:

1. High number of "multiple connections" in logs, e.g.
```
->[10.83.54.181:6783|3a:67:58:9e:ec:1a(ip-10-83-54-181.ec2.internal)]: connection shutting down due to error: Multiple connections to 3a:67:58:9e:ec:1a(ip-10-83-54-181.ec2.internal) added to da:d6:04:a2:f0:3f(ip-10-83-122-38.ec2.internal)
```
2. High number of "Inconsistent entries" in logs, e.g.
```
->[10.83.122.38:6783|da:d6:04:a2:f0:3f(ip-10-83-122-38.ec2.internal)]: connection shutting down due to error: Inconsistent entries for 100.103.128.0: owned by 7e:b0:95:48:d8:2d but incoming message says 9a:65:3c:8a:f3:c8
```
3. Existence of logs indicating that [nodes are fighting over the same IP range](https://github.com/weaveworks/weave/issues/3310), e.g.
```
->[10.83.122.38:20045|da:d6:04:a2:f0:3f(ip-10-83-122-38.ec2.internal)]: connection shutting down due to error: Received update for IP range I own at 100.101.224.0 v208: incoming message says owner 06:f2:22:1f:9a:ec v211
```
4. High number of "sleeve" related logs, even though we're configured for `fastdp`, e.g.
```
overlay_switch ->[da:d6:04:a2:f0:3f(ip-10-83-122-38.ec2.internal)] using sleeve
sleeve ->[10.83.97.1:6783|be:03:d4:4e:48:f5(ip-10-83-97-1.ec2.internal)]: Effective MTU verified at 8939
```
5. Consistent fluctuation of `pending` connections even when the cluster has stopped scaling.
6. Consistent connection terminations.
7. Inability of weave to clean up removed peers.

---

Some few hours later, you can see that while the cluster is in this state, weave consumes more resources, with memory gradually creeping up and up.

![image](https://user-images.githubusercontent.com/310235/62211105-fac59180-b39d-11e9-8f1d-de1ae40425d9.png)

---

To fix the inconsistent state we did the following: 
1. Terminated nodes that had nearly all of their peers unreachable.
2. Cleaned unreachable peers using `rmpeer`.
3. Performed a consistency of the IPAM ring across all nodes.
4. Discovered that there was a ["split-brain"](https://en.wikipedia.org/wiki/Split-brain_(computing)) in which there were two sources of truth for the IPAM ring.
5. Determined which set of nodes were maintaining the incorrect state.
5. Deleted the weave data & restarted weave pods for nodes with incorrect state (the number of nodes with incorrect state was about 1/3 of the cluster).

## How to reproduce it?

1. Scale up by some amount of nodes (the more nodes the more likely the issue will occur).
2. Scale down by the same amount of nodes.
3. Observe weave logs & metrics.

## Anything else we need to know?

Cloud provider: AWS, managed by Kops

<details><summary>Kubernetes Weave Manifest (click to expand)</summary>
<p>

```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    name: weave-net
  name: weave-net
  namespace: kube-system

---

apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  labels:
    name: weave-net
    role.kubernetes.io/networking: "1"
  name: weave-net
  namespace: kube-system
rules:
- apiGroups:
  - ""
  resources:
  - pods
  - namespaces
  - nodes
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - networking.k8s.io
  resources:
  - networkpolicies
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - ""
  resources:
  - nodes/status
  verbs:
  - patch
  - update

---

apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  labels:
    name: weave-net
    role.kubernetes.io/networking: "1"
  name: weave-net
  namespace: kube-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: weave-net
subjects:
- kind: ServiceAccount
  name: weave-net
  namespace: kube-system

---

apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  labels:
    name: weave-net
  name: weave-net
  namespace: kube-system
rules:
- apiGroups:
  - ""
  resourceNames:
  - weave-net
  resources:
  - configmaps
  verbs:
  - get
  - update
- apiGroups:
  - ""
  resources:
  - configmaps
  verbs:
  - create

---

apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  labels:
    name: weave-net
  name: weave-net
  namespace: kube-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: weave-net
subjects:
- kind: ServiceAccount
  name: weave-net
  namespace: kube-system

---

apiVersion: apps/v1
kind: DaemonSet
metadata:
  labels:
    name: weave-net
    role.kubernetes.io/networking: "1"
  name: weave-net
  namespace: kube-system
spec:
  minReadySeconds: 5
  selector:
    matchLabels:
      name: weave-net
      role.kubernetes.io/networking: "1"
  template:
    metadata:
      annotations:
        prometheus.io/scrape: "true"
        scheduler.alpha.kubernetes.io/critical-pod: ""
      labels:
        name: weave-net
        role.kubernetes.io/networking: "1"
    spec:
      containers:
      - command:
        - /home/weave/launch.sh
        env:
        - name: HOSTNAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
        - name: IPALLOC_RANGE
          value: 100.96.0.0/11
        - name: WEAVE_MTU
          value: "8912"
        image: weaveworks/weave-kube:2.5.1
        livenessProbe:
          httpGet:
            host: 127.0.0.1
            path: /status
            port: 6784
          initialDelaySeconds: 30
        name: weave
        ports:
        - name: metrics
          containerPort: 6782
        resources:
          limits:
            memory: 300Mi
          requests:
            cpu: 50m
            memory: 300Mi
        securityContext:
          privileged: true
        volumeMounts:
        - mountPath: /weavedb
          name: weavedb
        - mountPath: /host/opt
          name: cni-bin
        - mountPath: /host/home
          name: cni-bin2
        - mountPath: /host/etc
          name: cni-conf
        - mountPath: /host/var/lib/dbus
          name: dbus
        - mountPath: /lib/modules
          name: lib-modules
        - mountPath: /run/xtables.lock
          name: xtables-lock
      - args: []
        env:
        - name: HOSTNAME
          valueFrom:
            fieldRef:
              apiVersion: v1
              fieldPath: spec.nodeName
        image: weaveworks/weave-npc:2.5.1
        name: weave-npc
        ports:
        - name: metrics
          containerPort: 6781
        resources:
          limits:
            memory: 200Mi
          requests:
            cpu: 50m
            memory: 200Mi
        securityContext:
          privileged: true
        volumeMounts:
        - mountPath: /run/xtables.lock
          name: xtables-lock
      hostNetwork: true
      hostPID: true
      restartPolicy: Always
      securityContext:
        seLinuxOptions: {}
      serviceAccountName: weave-net
      tolerations:
      - effect: NoSchedule
        operator: Exists
      - key: CriticalAddonsOnly
        operator: Exists
      volumes:
      - hostPath:
          path: /var/lib/weave
        name: weavedb
      - hostPath:
          path: /opt
        name: cni-bin
      - hostPath:
          path: /home
        name: cni-bin2
      - hostPath:
          path: /etc
        name: cni-conf
      - hostPath:
          path: /var/lib/dbus
        name: dbus
      - hostPath:
          path: /lib/modules
        name: lib-modules
      - hostPath:
          path: /run/xtables.lock
          type: FileOrCreate
        name: xtables-lock
  updateStrategy:
    type: RollingUpdate
```
</p>
</details>

```
Weave MTU: 8912
```

## Versions:
```console
/home/weave # ./weave --local version
weave 2.5.2
```
```console
root@ip-10-84-96-97:/home/admin# docker version
Client:
 Version:           18.06.3-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        d7080c1
 Built:             Wed Feb 20 02:28:26 2019
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.3-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       d7080c1
  Built:            Wed Feb 20 02:26:51 2019
  OS/Arch:          linux/amd64
  Experimental:     false
```
```console
root@ip-10-84-96-97:/home/admin# docker info
Containers: 23
 Running: 23
 Paused: 0
 Stopped: 0
Images: 21
Server Version: 18.06.3-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: a592beb5bc4c4092b1b1bac971afed27687340c5
init version: fec3683
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.9.0-9-amd64
Operating System: Debian GNU/Linux 9 (stretch)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.799GiB
Name: ip-10-84-96-97
ID: PJLG:ZCPD:TSFR:C3PT:R3MO:ALO7:6TKF:3BWR:HL7D:ICLQ:W2DA:3IMT
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```
```console
root@ip-10-84-96-97:/home/admin# uname -a
Linux ip-10-84-96-97 4.9.0-9-amd64 #1 SMP Debian 4.9.168-1 (2019-04-12) x86_64 GNU/Linux
```
```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.9", GitCommit:"e09f5c40b55c91f681a46ee17f9bc447eeacee57", GitTreeState:"clean", BuildDate:"2019-05-27T16:08:57Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"12", GitVersion:"v1.12.8", GitCommit:"a89f8c11a5f4f132503edbc4918c98518fd504e3", GitTreeState:"clean", BuildDate:"2019-04-23T04:41:47Z", GoVersion:"go1.10.8", Compiler:"gc", Platform:"linux/amd64"}
```
```console
$ kops version
Version 1.12.2
```

## Logs

We don't currently have the weave logs available right now, but we hope to make them available soon.

If you'd like me to search for specific things in the logs in the meantime I can happily provide snippets.

## Diagnostics

Weave diagnostics: [weave-sandbox-20190729.tar.gz](https://github.com/weaveworks/weave/files/3451881/weave-sandbox-20190729.tar.gz)

In the above you will find the following for every weave pod in the cluster a few hours after the cluster become inconsistent:
* Connections
* Peers
* Targets
* IPAM
* Report
* Heap Dump

Also includes the weave config map at the root.

**Comment by murali-reddy** [comment]
@zacblazic thanks for reporting the issue.

I just did some scaling tests scale up 50 to 100 nodes and scale down from 100 to 50 several times. My observation is 2.5.2 is working perfectly fine. I also tested with 2.5.0 on which scale down certainly showed the symptoms you described. 

2.5.1 and 2.5.2 releases have fixes (ensure rmpeer of deleted nodes and resolving IPAM conflix isssues) that address some of the behaviour you reported in this issue.

> The cluster was scaled up from 62 to 102 nodes. The scale up was handled well and apart from a few logs that indicated we were in "sleeve" mode and that we had "multiple connections", however those logs only happened during startup of weave on the new nodes. Next, the cluster was scaled back down to 62 nodes, which is where the trouble began from what I can tell:

Did you had to scale up and down several times to see this behaviour or was it just at first scale down?

**Comment by mmerrill3** [comment]
See #3722 and the PR is #3724.  It's probably the same thing

**Comment by bboreham** [comment]
I'm going to close this as fixed in 2.6.0 - thanks to @mmerrill3 for the fix.

@zacblazic if it re-occurs feel free to reopen and give us some more details.

**Comment by alok87** [comment]
@zacblazic We are facing the same issue. We upgraded to `2.6.0` but the current inconsistency is still there. Can you please share more details on how did you fix this mess.
